daemon/probes.go

```diff
@@ -0,0 +1,290 @@
+			return
+		}
+	}
+
+	return
+}
+
+// copyFile copies the file from source to destination. Before calling this
+// function, there should be no file under destination path.
+func copyFile(src, dest string) error {
+	in, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer in.Close()
+
+	out, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	_, err = io.Copy(out, in)
+	return err
+}
+
+// probeRunLl runs the probe for compiling the program with clang. It returns
+// number of bytes written to infoFile, number of bytes written to warningFile
+// and error.
+func probeRunLl(featureFile, infoFile io.Writer, probesDir, libIncludeDir, outDir string) (infoFileBytes int, err error) {
+	outFilename := path.Join(outDir, "raw_probe.t")
+
+	err = filepath.Walk(probesDir, func(probePath string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if strings.HasSuffix(info.Name(), ".t") {
+			outBin := path.Join(outDir, path.Base(probePath))
+			if err := copyFile(probePath, outFilename); err != nil {
+				return err
+			}
+			out, err := exec.Command(
+				"clang",
+				"-O2",
+				"-Wall",
+				fmt.Sprintf("-I%s", probesDir),
+				fmt.Sprintf("-I%s", libIncludeDir),
+				fmt.Sprintf("-I%s", outDir),
+				path.Join(probesDir, "raw_main.c"),
+				"-o",
+				outBin,
+			).CombinedOutput()
+			if err != nil {
+				return fmt.Errorf("clang failed: %s: %s", err, out)
+			}
+
+			var stdoutBuf, stderrBuf bytes.Buffer
+			cmd := exec.Command(outBin)
+			cmd.Stdout = &stdoutBuf
+			cmd.Stderr = &stderrBuf
+			if err := cmd.Run(); err != nil {
+				return err
+			}
+
+			if _, err := stdoutBuf.WriteTo(featureFile); err != nil {
+				return err
+			}
+			n, err := stderrBuf.WriteTo(infoFile)
+			if err != nil {
+				return err
+			}
+			infoFileBytes += int(n)
+		}
+		return nil
+	})
+	return
+}
+
+// writeFeatures wrtites definitions of features into the header file which is
+// going to be used by BPF programs.
+func writeFeatures(featureFile, infoFile, warningFile io.Writer, probesDir, libIncludeDir, outDir string) (infoFileBytes int, warningFileBytes int, err error) {
+	io.WriteString(featureFile, `#ifndef BPF_FEATURES_H_
+#define BPF_FEATURES_H_
+`)
+
+	var ni, nw int
+	ni, nw, err = probeKernelConfig(infoFile, warningFile)
+	if err != nil {
+		err := fmt.Errorf("kernel config probe failed: %s", err)
+		return infoFileBytes, warningFileBytes, err
+	}
+	infoFileBytes += ni
+	warningFileBytes += nw
+
+	var n int
+	n, err = probeRunLl(featureFile, infoFile, probesDir, libIncludeDir, outDir)
+	if err != nil {
+		err := fmt.Errorf("ll probe failed: %s", err)
+		return infoFileBytes, warningFileBytes, err
+	}
+	infoFileBytes += n
+
+	io.WriteString(featureFile, "#endif /* BPF_FEATURES_H_ */\n")
+
+	return
+}
+
+// runProbes runs probes for BPF features. It generates a header file for BPF
+// programs and log files.
+func runProbes(bpfDir, stateDir string) error {
```

Comment:
Most of the code in this function seems to be the same operation for two or three files (rm, open, close, sync), but copied and pasted. I wonder if it would be a little easier to ensure that they are exactly the same by, for example, refactoring some of the common sets of operations.

Reply:
I tried my best. I hope that the `fileWithCleanup` structure is abstract enough.